### PR TITLE
Position control with Ignition Fortress

### PR DESCRIPTION
**Pull request by benediktkreis**
Hi,

this is my first contribution to ROS2. So please forgive me if I have made any mistakes.

I have tried to run `ros2 launch ur_simulation_ignition ur_sim_moveit.launch.py`, but it did not work. The following command is working in my setup with the following code changes.

### My setup
ROS2 version: Galactic
Ignition version: Fortress
Moveit2 version: Binary install for ROS2 Galactic

### Command
Example command to launch Moveit and Ignition with UR5e model: 
`ros2 launch ur_simulation_ignition ur_sim_moveit.launch.py ur_type:=ur5e`

### Code changes
-  [ur_moveit.launch.py](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/main/ur_bringup/launch/ur_moveit.launch.py)  was added from [ur_bringup](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/tree/main/ur_bringup), because there is no launch file with this name in the [ur_moveit_config](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/tree/main/ur_moveit_config) and to set use_sim_time=True in the parameters of the move_group_node. I am not sure why this package was changed in [#1](https://github.com/UniversalRobots/Universal_Robots_ROS2_Ignition_Simulation/pull/1)
- update_rate in [ur_controllers.yaml](https://github.com/UniversalRobots/Universal_Robots_ROS2_Ignition_Simulation/blob/ros2/ur_simulation_ignition/config/ur_controllers.yaml) was set to 1000 Hz to get rid of Ignition warning: "Desired controller update period (0.002 s) is slower than the gazebo simulation period (0.001 s).".
- Position control with ign_ros2_control only works with the [galactic branch](https://github.com/ignitionrobotics/ign_ros2_control/tree/galactic) (see bug fix [#29](https://github.com/ignitionrobotics/ign_ros2_control/pull/29)).
- delay_rviz_after_joint_state_broadcaster_spawner in [ur_sim_control.launch.py](https://github.com/UniversalRobots/Universal_Robots_ROS2_Ignition_Simulation/blob/ros2/ur_simulation_ignition/launch/ur_sim_control.launch.py) caused the launching of 2 RVIZ windows. So I have removed it.

### Output
UR5e_Ignition_Moveit

### CODE REVIEW

File: ur_simulation_ignition/launch/ur_sim_control.launch.py [hunk 1]
```diff
@@ -149,14 +149,6 @@ def launch_setup(context, *args, **kwargs):
         arguments=["joint_state_broadcaster", "--controller-manager", "/controller_manager"],
     )
 
-    # Delay rviz start after `joint_state_broadcaster`
-    delay_rviz_after_joint_state_broadcaster_spawner = RegisterEventHandler(
-        event_handler=OnProcessExit(
-            target_action=joint_state_broadcaster_spawner,
-            on_exit=[rviz_node],
-        )
-    )
```
Comment (destogl) [inline ur_simulation_ignition/launch/ur_sim_control.launch.py]: Can we please not remove this but add a condition instead `condition=IfCondition(launch_rviz),`?

File: ur_simulation_ignition/launch/ur_moveit.launch.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
+# Copyright (c) 2021 PickNik, Inc.
```
Comment (destogl) [inline ur_simulation_ignition/launch/ur_moveit.launch.py:1]: We should use file from the "Driver" repository. It will get merged here: UniversalRobots/Universal_Robots_ROS2_Driver#278

It would be great if you could also test this repos with that PR and give us a feedback there if it worked for you.

File: Universal_Robots_ROS2_Ignition_Simulation-not-released.galactic.repos [hunk 3]
```diff
@@ -3,7 +3,7 @@ repositories:
   ign_ros2_control:
     type: git
     url: https://github.com/ignitionrobotics/ign_ros2_control.git
-    version: main
```
Comment (destogl) [inline Universal_Robots_ROS2_Ignition_Simulation-not-released.galactic.repos:6]: Thanks!

File: ur_simulation_ignition/launch/ur_sim_control.launch.py [hunk 4]
```diff
@@ -186,6 +178,17 @@ def launch_setup(context, *args, **kwargs):
         ],
     )
 
+    # ros_ign_bridge (clock -> ROS 2)
```
Comment (destogl) [inline ur_simulation_ignition/launch/ur_sim_control.launch.py:190]: Why is this needed?

File: ur_simulation_ignition/launch/ur_sim_moveit.launch.py [hunk 5]
```diff
@@ -67,10 +67,11 @@ def launch_setup(context, *args, **kwargs):
 
     ur_moveit_launch = IncludeLaunchDescription(
         PythonLaunchDescriptionSource(
-            [FindPackageShare("ur_moveit_config"), "/launch", "/ur_moveit.launch.py"]
+            [FindPackageShare("ur_simulation_ignition"), "/launch", "/ur_moveit.launch.py"]
```
Comment (destogl) [inline ur_simulation_ignition/launch/ur_sim_moveit.launch.py:70]: Please test with the above-mentioned branch.

**Review by destogl — COMMENTED** [review]
Thanks for testing this repository and for your contribution!

We are currently having a PR with a changed MoveIt file in the "Driver" repository open. Sorry I was too fast with merging changes when I was enabling MoveIt. I testing this directly with the mentioned branch.

To have Ignition correctly working with MoveIt I had to fix the [following in MoveIt2](https://github.com/ros-planning/moveit2/pull/1014).

I am not sure why is this working for you at all. Nevertheless, it is great first step to make this repo nice and clean. Please consider my comments below.

File: ur_simulation_ignition/launch/ur_sim_control.launch.py (hunk 1 above)
Comment (benediktkreis) [reply ur_simulation_ignition/launch/ur_sim_control.launch.py]: Sure, this does the trick. I've updated the launch file accordingly.

File: ur_simulation_ignition/launch/ur_moveit.launch.py (hunk 2 above)
Comment (benediktkreis) [reply ur_simulation_ignition/launch/ur_moveit.launch.py:1]: I've tested the update-simplify-moveit-file branch. In general it works, but I had to do some modifications, because it doesn't work with the scaled_joint_trajectory_controller. The joint_trajectory_controller has to be set as default in the [controllers.yaml](https://github.com/benediktkreis/Universal_Robots_ROS2_Ignition_Simulation/blob/sim_time/ur_simulation_ignition/config/controllers.yaml). I guess the scaled one is not supported by the [ign_ros2_control](https://github.com/ignitionrobotics/ign_ros2_control) plugin. That is why I've copied the [ur_moveit.launch.py](https://github.com/benediktkreis/Universal_Robots_ROS2_Ignition_Simulation/blob/sim_time/ur_simulation_ignition/launch/ur_moveit.launch.py) and the [controllers.yaml](https://github.com/benediktkreis/Universal_Robots_ROS2_Ignition_Simulation/blob/sim_time/ur_simulation_ignition/config/controllers.yaml) with the one from your branch and modified them accordingly. Furthermore, without the robot_description_planning in [ur_moveit.launch.py](https://github.com/benediktkreis/Universal_Robots_ROS2_Ignition_Simulation/blob/sim_time/ur_simulation_ignition/launch/ur_moveit.launch.py) it isn't possible to plan a path using the drag function ("virtual ball" at the end effector) in RVIZ, I could just plan using predefined goal states like "home", "up" and "test_configuration".

Hint: I didn't use your [Moveit2 branch](https://github.com/destogl/moveit2/tree/fix-controller-check-for-simulation) [#1014](https://github.com/ros-planning/moveit2/pull/1014), because I had problems building it. That's why I've sticked to the binary version for Galactic.

File: ur_simulation_ignition/launch/ur_sim_control.launch.py (hunk 4 above)
Comment (benediktkreis) [reply ur_simulation_ignition/launch/ur_sim_control.launch.py:190]: Otherwise the /clock topic isn't populated which is used e.g. by the robot_state_publisher (see attached screenshot from rqt_graph) resulting in an error message such as this one: 
`[rviz2-9] [WARN] [1644917311.649767236] [moveit_ros.planning_scene_monitor.planning_scene_monitor]: Maybe failed to update robot state, time diff: 1644917299.652s`

rqt_graph_clock_cropped

The solution is inspired by [Andrej Orsula](https://github.com/AndrejOrsula)'s repository [panda_ign_moveit2](https://github.com/AndrejOrsula/panda_ign_moveit2/blob/master/panda_moveit_config/launch/ex_ign_control.launch.py)

File: ur_simulation_ignition/launch/ur_sim_moveit.launch.py (hunk 5 above)
Comment (benediktkreis) [reply ur_simulation_ignition/launch/ur_sim_moveit.launch.py:70]: Has been tested, see other comment.

File: Universal_Robots_ROS2_Ignition_Simulation-not-released.galactic.repos (hunk 3 above)
Comment (benediktkreis) [reply Universal_Robots_ROS2_Ignition_Simulation-not-released.galactic.repos:6]: No problem :)

**Comment by benediktkreis** [comment]
Hi @destogl, what do you think about my last commits and comments? Are further adjustments required?

**Comment by fmauch** [comment]
Hi, we haven't been giving this repository a lot of attention recently. We are currently in the process of making things work again. Currently, Humble and Iron (Provided the description is built from source ATM) should be working, the version from #7 should work on Rolling. All tested with the default GZ version.

Therefore, I would go ahead and close this, as I guess this would be included / supersed by the latest changes. Feel free to comment / reopen if you think otherwise.
